modules/iothubdeviceping/src/iothubdeviceping.c

```diff
@@ -0,0 +1,738 @@
+        }
+        else
+        {
+            LogError("config->EH_HOST malloc returned NULL\r\n");
+        }
+        if ((clonedConfig->EH_KEY_NAME = (const char *)malloc(strlen(config->EH_KEY_NAME) + 1)) != NULL)
+        {
+            memcpy((char *)clonedConfig->EH_KEY_NAME, (char *)config->EH_KEY_NAME, strlen(config->EH_KEY_NAME) + 1);
+        }
+        else
+        {
+            LogError("config->EH_KEY_NAME malloc returned NULL\r\n");
+        }
+        if ((clonedConfig->EH_KEY = (const char *)malloc(strlen(config->EH_KEY) + 1)) != NULL)
+        {
+            memcpy((char *)clonedConfig->EH_KEY, (char *)config->EH_KEY, strlen(config->EH_KEY) + 1);
+        }
+        else
+        {
+            LogError("config->EH_KEY malloc returned NULL\r\n");
+        }
+        if ((clonedConfig->EH_COMP_NAME = (const char *)malloc(strlen(config->EH_COMP_NAME) + 1)) != NULL)
+        {
+            memcpy((char *)clonedConfig->EH_COMP_NAME, (char *)config->EH_COMP_NAME, strlen(config->EH_COMP_NAME) + 1);
+        }
+        else
+        {
+            LogError("config->EH_COMP_NAME malloc returned NULL\r\n");
+        }
+        if ((clonedConfig->EH_PARTITION_NUM = (const char *)malloc(strlen(config->EH_PARTITION_NUM) + 1)) != NULL)
+        {
+            memcpy((char *)clonedConfig->EH_PARTITION_NUM, (char *)config->EH_PARTITION_NUM, strlen(config->EH_PARTITION_NUM) + 1);
+        }
+        else
+        {
+            LogError("config->EH_PARTITION_NUM malloc returned NULL\r\n");
+        }
+    }
+    return clonedConfig;
+}
+
+static MODULE_HANDLE IoTHubDevicePing_Create(BROKER_HANDLE brokerHandle, const void *configuration)
+{
+    // set g_echoreply false
+    g_echoreply = false;
+    IOTHUBDEVICEPING_HANDLE_DATA *result = NULL;
+    if (
+        (brokerHandle == NULL) ||
+        (configuration == NULL) ||
+        (((const IOTHUBDEVICEPING_CONFIG *)configuration)->DeviceConnectionString == "NULL") ||
+        (((const IOTHUBDEVICEPING_CONFIG *)configuration)->EH_HOST == "NULL") ||
+        (((const IOTHUBDEVICEPING_CONFIG *)configuration)->EH_KEY_NAME == "NULL") ||
+        (((const IOTHUBDEVICEPING_CONFIG *)configuration)->EH_KEY == "NULL") ||
+        (((const IOTHUBDEVICEPING_CONFIG *)configuration)->EH_COMP_NAME == "NULL"))
+    {
+        LogError("invalid arg brokerHandle=%p, configuration=%p\r\n", brokerHandle, configuration);
+        result = NULL;
+    }
+    else
+    {
+        result = malloc(sizeof(IOTHUBDEVICEPING_HANDLE_DATA));
+        if (result == NULL)
+        {
+            LogError("malloc returned NULL\r\n");
+        }
+        else
+        {
+            result->config = (IOTHUBDEVICEPING_CONFIG *)clone_iothub_device_ping_config(configuration);
+            result->iotHubClientHandle = IoTHubClient_LL_CreateFromConnectionString(((const IOTHUBDEVICEPING_CONFIG *)configuration)->DeviceConnectionString, HTTP_Protocol);
```

Comment:
Hi @damonbarry, thanks again for the great feedback. I've just added support for sending over AMQP and MQTT as suggested. we want the user to easily verify the correctness of their initial setup of the gateway environment as well as reachability of iot hub they've provisioned. Being able to send over the three protocols and confirm the arrival of the message with ease is useful to users. Please review. Thanks.